### ISSUE TITLE
twisted.test.test_pb.NewCredTestCase.test_clientConnectionLost failing on windows

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2876]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#2876|
|Type|defect|
|Created|2007-11-07 14:33:13Z|
|Branch|https://github.com/twisted/twisted/tree/test-pb-reconnecting-2876|

Windows slave configuration changed slightly (maybe a red herring).  ```test_clientConnectionLost``` has errors now, on several Windows slaves:

```
===============================================================================
[ERROR]: twisted.test.test_pb.NewCredTestCase.test_clientConnectionLost

Traceback (most recent call last):
Failure: twisted.trial.util.DirtyReactorAggregateError: Reactor was unclean.
DelayedCalls: (set twisted.internet.base.DelayedCall.debug = True to debug)
<DelayedCall 22763960 [29.9850001335s] called=0 cancelled=0 Client.failIfNotConnected(TimeoutError('',))>
===============================================================================
[ERROR]: twisted.test.test_pb.NewCredTestCase.test_clientConnectionLost

Traceback (most recent call last):
Failure: twisted.trial.util.DirtyReactorAggregateError: Reactor was unclean.
Selectables:
<<class 'twisted.internet.tcp.Client'> to ('127.0.0.1', 4176) at 32904f0>
-------------------------------------------------------------------------------
```

This happens on three builders covering two different versions of Windows and two different versions of Python.

Attachments:

* [2876.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/3bf/3bfe375dd1f702aa715048b239063f191a619583/77e42f68e43b2540ca5daed7f4a8e8fd759e582a.diff) (1539 bytes) - added by therve on 2007-12-17 12:32:09Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__2876 2876
type__defect defect
reporter__exarkun exarkun
priority__highest highest
milestone__ 
branch__branches_test_pb_reconnecting_2876 branches/test-pb-reconnecting-2876
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__win32 win32
time__1194445993000000 1194445993000000
changetime__1205257618000000 1205257618000000
version__None None
owner__ 
cc__therve
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_1"></a>|@therve commented|
|-|-|

Note that these are unclean errors, which probably explain why they didn't appear before.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

Good point.  I'm going to change the buildbot configuration so unclean errors are only warnings on the windows builders.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_3"></a>|@therve removed owner|
|-|-|

The problem was that the factory was trying to reconnect even on the second disconnection. I also use assertIsInstance where necessary, and correct an error on the second test (the wrong object was checked). Please review.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_4"></a>|@therve commented|
|-|-|

(In [22793]) Branching to 'test-pb-reconnecting-2876'

**Comment by twisted-trac** [comment]
|<img alt="bigdog's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_5"></a>|bigdog set owner to @therve|
|-|-|

I verified that the branch does no harm.  The code seems reasonable and is covered by current test cases.

I ran the branch on windows and unix with no negative impact.

This code should be ready to go

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_7"></a>|@therve set status to `closed`|
|-|-|

(In [22796]) Merge test-pb-reconnecting-2876

Author: therve
Reviewer: bigdog
Fixes [#2876](https://github.com/twisted/twisted/issues/2876)

Fix unclean errors in
twisted.test.test_pb.NewCredTestCase.test_clientConnectionLost under win32.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
